apps/opencs/view/render/terrainshapemode.cpp

```diff
@@ -0,0 +1,1459 @@
+                if (landShapeNew[col * ESM::Land::LAND_SIZE + row] <= 0) floatLodHeight = landShapeNew[col * ESM::Land::LAND_SIZE + row] / 16;
+                if (floatLodHeight > std::numeric_limits<signed char>::max()) lodHeight = std::numeric_limits<signed char>::max();
+                else if (floatLodHeight < std::numeric_limits<signed char>::min()) lodHeight = std::numeric_limits<signed char>::min();
+                else lodHeight = static_cast<signed char>(floatLodHeight);
+                mapLodShapeNew[j * sqrtLandGlobalMapLodSize + i] = lodHeight;
+            }
+        }
+        pushEditToCommand(landShapeNew, document, landTable, cellId);
+        pushLodToCommand(mapLodShapeNew, document, landTable, cellId);
+    }
+
+    for(CSMWorld::CellCoordinates cellCoordinates: mAlteredCells)
+    {
+        std::string cellId = CSMWorld::CellCoordinates::generateId(cellCoordinates.getX(), cellCoordinates.getY());
+        const CSMWorld::LandHeightsColumn::DataType landShapePointer = landTable.data(landTable.getModelIndex(cellId, landshapeColumn)).value<CSMWorld::LandHeightsColumn::DataType>();
+        const CSMWorld::LandHeightsColumn::DataType landRightShapePointer = landTable.data(landTable.getModelIndex(CSMWorld::CellCoordinates::generateId(cellCoordinates.getX() + 1, cellCoordinates.getY()), landshapeColumn)).value<CSMWorld::LandHeightsColumn::DataType>();
+        const CSMWorld::LandHeightsColumn::DataType landDownShapePointer = landTable.data(landTable.getModelIndex(CSMWorld::CellCoordinates::generateId(cellCoordinates.getX(), cellCoordinates.getY() + 1), landshapeColumn)).value<CSMWorld::LandHeightsColumn::DataType>();
+        const CSMWorld::LandNormalsColumn::DataType landNormalsPointer = landTable.data(landTable.getModelIndex(cellId, landnormalsColumn)).value<CSMWorld::LandNormalsColumn::DataType>();
+        CSMWorld::LandNormalsColumn::DataType landNormalsNew(landNormalsPointer);
+
+        // Generate land normals record
+        for(int i = 0; i < ESM::Land::LAND_SIZE; ++i)
+        {
+            for(int j = 0; j < ESM::Land::LAND_SIZE; ++j)
+            {
+                osg::Vec3f v1;
+                osg::Vec3f v2;
+                osg::Vec3f normal;
+                float hyp;
+
+                v1.x() = 128;
+                v1.y() = 0;
+                if (i < ESM::Land::LAND_SIZE - 1) v1.z() = landShapePointer[j * ESM::Land::LAND_SIZE + i + 1] - landShapePointer[j * ESM::Land::LAND_SIZE + i];
+                else
+                {
+                    std::string cellId = CSMWorld::CellCoordinates::generateId(cellCoordinates.getX() + 1, cellCoordinates.getY());
+                    if (!noCell(cellId) && !noLand(cellId) && noLandLoaded(cellId))
+                        v1.z() = landRightShapePointer[j * ESM::Land::LAND_SIZE + 1] - landShapePointer[j * ESM::Land::LAND_SIZE + i];
+                    else
+                        v1.z() = 0;
```

Comment:
This branch is redundant. By default `osg::Vec3f` initializes all components with zeroes.

Reply:
Should also be `!noLandLoaded(cellId)` ... fixed that too.